File: factorize-gaussian-integer-javascript/FactorizeGaussianInteger.java
/* 
 * Factorize Gaussian integer (Java)
 * 
 * Run this program with a single argument containing the complex number to be factorized. Examples:
 * - Command line: java FactorizeGaussianInteger "-1"
 *   Output: (-1)
 * - Command line: java FactorizeGaussianInteger "2"
 *   Output: (-1i)(1 + 1i)(1 + 1i)
 * - Command line: java FactorizeGaussianInteger "7"
 *   Output: (7)
 * - Command line: java FactorizeGaussianInteger "5 + 9i"
 *   Output: (1 + 1i)(7 + 2i)
 * 
 * Copyright (c) 2016 Project Nayuki
 * All rights reserved. Contact Nayuki for licensing.
 * https://www.nayuki.io/page/factorize-gaussian-integer-javascript
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public final class FactorizeGaussianInteger {
	
	public static void main(String[] args) {
		GaussianInteger num = new GaussianInteger(args[0]);
		List<GaussianInteger> factorization = num.factorize();
		for (GaussianInteger factor : factorization)
			System.out.print("(" + factor.toString() + ")");
		System.out.println();
	}
	
}



// Immutable complex number where real and imag are integers, and (0 <= real, imag < 2^31).
final class GaussianInteger {
	
	/* Fields */
	
	public final int real;
	public final int imag;
	
	
	
	/* Constructors */
	
	public GaussianInteger(int real, int imag) {
		if (real == Integer.MIN_VALUE || imag == Integer.MIN_VALUE)
			throw new IllegalArgumentException("Value out of range");
		this.real = real;
		this.imag = imag;
	}
	
	
	private GaussianInteger(long real, long imag) {
		if (real <= Integer.MIN_VALUE || real > Integer.MAX_VALUE ||
		    imag <= Integer.MIN_VALUE || imag > Integer.MAX_VALUE)
			throw new IllegalArgumentException("Value out of range");
		this.real = (int)real;
		this.imag = (int)imag;
	}
	
	
	public GaussianInteger(String s) {
		// Spaces are not allowed between digits. Otherwise remove all whitespace as preprocessing.
		if (WHITESPACE_BETWEEN_DIGITS_REGEX.matcher(s).matches())
			throw new IllegalArgumentException("Invalid number");
		s = WHITESPACE_REGEX.matcher(s).replaceAll("");
		
		// Match one of the syntax cases
		Matcher m;
		if ((m = REAL_REGEX.matcher(s)).matches()) {  // e.g. 1, +0, -2
			real = Integer.parseInt(m.group(1));
			imag = 0;
		} else if ((m = IMAG_REGEX.matcher(s)).matches()) {  // e.g. i, 4i, -3i
			real = 0;
			imag = Integer.parseInt(m.group(1) + (m.group(2).equals("") ? "1" : m.group(2)));
		} else if ((m = REAL_IMAG_REGEX.matcher(s)).matches()) {  // e.g. 1+2i, -3-4i, +5+i
			real = Integer.parseInt(m.group(1));
			imag = Integer.parseInt(m.group(2) + (m.group(3).equals("") ? "1" : m.group(3)));
		} else if ((m = IMAG_REAL_REGEX.matcher(s)).matches()) {  // e.g. 2i+1, -4i-3, +i+5
			real = Integer.parseInt(m.group(3));
			imag = Integer.parseInt(m.group(1) + (m.group(2).equals("") ? "1" : m.group(2)));
		} else
			throw new IllegalArgumentException("Invalid number");
		
		if (real == Integer.MIN_VALUE || imag == Integer.MIN_VALUE)
			throw new IllegalArgumentException("Value out of range");
	}
	
	
	
	/* Methods */
	
	public long norm() {
		return (long)real * real + (long)imag * imag;  // Does not overflow
	}
	
	
	public GaussianInteger multiply(GaussianInteger other) {
		long re = (long)real * other.real - (long)imag * other.imag;  // Does not overflow
		long im = (long)imag * other.real + (long)real * other.imag;  // Does not overflow
		try {
			return new GaussianInteger(re, im);
		} catch (IllegalArgumentException e) {
			throw new ArithmeticException("Overflow");
		}
	}
	
	
	public boolean isDivisibleBy(int re, int im) {
		long divisorNorm = (long)re * re + (long)im * im;
		return ((long) real * re + (long)imag * im) % divisorNorm == 0 &&
		       ((long)-real * im + (long)imag * re) % divisorNorm == 0;
	}
	
	
	public GaussianInteger divide(GaussianInteger other) {
		if (!isDivisibleBy(other.real, other.imag))
			throw new ArithmeticException("Cannot divide");
		long re = ((long) real * other.real + (long)imag * other.imag) / other.norm();
		long im = ((long)-real * other.imag + (long)imag * other.real) / other.norm();
		try {
			return new GaussianInteger(re, im);
		} catch (IllegalArgumentException e) {
			throw new AssertionError(e);
		}
	}
	
	
	public List<GaussianInteger> factorize() {
		List<GaussianInteger> result = new ArrayList<>();
		if (norm() <= 1) {  // 0, 1, -1, i, -i
			result.add(this);
			return result;
		}
		
		GaussianInteger temp = this;
		GaussianInteger check = new GaussianInteger(1, 0);
		while (temp.norm() > 1) {
			GaussianInteger factor = temp.findPrimeFactor();
			result.add(factor);
			temp = temp.divide(factor);
			check = check.multiply(factor);
		}
		check = check.multiply(temp);
		if (temp.norm() != 1 || check.real != real || check.imag != imag)
			throw new AssertionError();
		if (temp.real != 1)  // -1, i, -i
			result.add(temp);
		
		Collections.sort(result, new Comparator<GaussianInteger>() {
			public int compare(GaussianInteger x, GaussianInteger y) {
				if      (x.norm() < y.norm()) return -1;
				else if (x.norm() > y.norm()) return +1;
				else if (x.real > y.real) return -1;
				else if (x.real < y.real) return +1;
				else return 0;
			}
		});
		return result;
	}
	
	
	private GaussianInteger findPrimeFactor() {
		long norm = norm();
		if (norm % 2 == 0)
			return new GaussianInteger(1, 1);
		
		for (long i = 3, end = sqrt(norm); i <= end; i += 2) {  // Find factors of norm
			if (norm % i == 0) {
				if (i % 4 == 3)
					return new GaussianInteger(i, 0);
				else {
					for (long re = sqrt(i); re > 0; re--) {
						long im = sqrt(i - re * re);
						if (re * re + im * im == i) {
							if (re > Integer.MAX_VALUE || im > Integer.MAX_VALUE)
								throw new ArithmeticException("Overflow");
							else if (isDivisibleBy((int)re, (int)im))
								return new GaussianInteger(re, im);
						}
					}
				}
			}
		}
		
		// This number itself is prime. Rotate so that the argument is in [0, pi/2)
		GaussianInteger temp = this;
		while (temp.real < 0 || temp.imag < 0)
			temp = temp.multiply(new GaussianInteger(0, 1));
		return temp;
	}
	
	
	public String toString() {
		if (imag == 0)
			return real + "";
		else if (real == 0)
			return imag + "i";
		else if (real != 0 && imag > 0)
			return real + " + " + imag + "i";
		else if (real != 0 && imag < 0)
			return real + " - " + (-imag) + "i";
		else
			throw new AssertionError();
	}
	
	
	
	/* Static members */
	
	private static long sqrt(long x) {
		if (x < 0)
			throw new IllegalArgumentException("Square root of negative number");
		long y = 0;
		for (long i = 1L << 31; i != 0; i >>>= 1) {
			y |= i;
			if (y > 3037000499L || y * y > x)
				y ^= i;
		}
		return y;
	}
	
	
	private static final Pattern WHITESPACE_BETWEEN_DIGITS_REGEX = Pattern.compile("\\d\\s+\\d");
	private static final Pattern WHITESPACE_REGEX = Pattern.compile("\\s+");
	
	private static final Pattern REAL_REGEX = Pattern.compile("([+-]?\\d+)");
	private static final Pattern IMAG_REGEX = Pattern.compile("([+-]?)(\\d*)i");
	private static final Pattern REAL_IMAG_REGEX = Pattern.compile("([+-]?\\d+)([+-])(\\d*)i");
	private static final Pattern IMAG_REAL_REGEX = Pattern.compile("([+-]?)(\\d*)i([+-]\\d+)");
	
}

File: factorize-gaussian-integer-javascript/factorize-gaussian-integer.ts
/* 
 * Factorize Gaussian integer
 * 
 * Copyright (c) 2018 Project Nayuki
 * All rights reserved. Contact Nayuki for licensing.
 * https://www.nayuki.io/page/factorize-gaussian-integer-javascript
 */

"use strict";


namespace app {
	
	/* 
	 * Handles the HTML input/output for factoring a Gaussian integer.
	 */
	export function doFactor(ev: Event|null): void {
		if (ev !== null)
			ev.preventDefault();
		
		let outElem = document.getElementById("factorization") as HTMLElement;
		while (outElem.firstChild != null)
			outElem.removeChild(outElem.firstChild);
		const input = (document.getElementById("number") as HTMLInputElement).value;
		if (/^\s*$/.test(input)) {  // Blank input
			outElem.textContent = NBSP;
			return;
		}
		
		// DOM helper function
		function appendTextNode(elem: HTMLElement, s: string): void {
			elem.appendChild(document.createTextNode(s));
		}
		
		// Formatting helper function
		function appendGaussianInteger(n: GaussianInteger): void {
			const s: string = n.toString();
			if (s.charAt(s.length - 1) != "i")
				appendTextNode(outElem, s);
			else {
				let varElem = document.createElement("var");
				varElem.textContent = "i";
				appendTextNode(outElem, s.substring(0, s.length - 1));
				outElem.appendChild(varElem);
			}
		}
		
		try {
			const num: GaussianInteger = GaussianInteger.parseString(input);
			const factorization: Array<GaussianInteger> = num.factorize();
			
			appendGaussianInteger(num);
			appendTextNode(outElem, " = ");
			factorization.forEach((factor, i) => {
				if (i > 0)
					appendTextNode(outElem, " ");
				appendTextNode(outElem, "(");
				appendGaussianInteger(factor);
				appendTextNode(outElem, ")");
			});
		} catch (e) {
			outElem.appendChild(document.createTextNode(e.toString()));
		}
	}
	
	
	export function doRandom(): void {
		function randInt(): number {
			return Math.floor(Math.random() * 2000) - 1000;
		}
		
		const type: number = Math.random();
		let str: string;
		if (type < 0.2)
			str = randInt().toString();
		else if (type < 0.3)
			str = randInt() + "i";
		else {
			const real = randInt();
			const imag = randInt();
			str = real + (imag >= 0 ? " + " : " - ") + Math.abs(imag) + "i";
		}
		(document.getElementById("number") as HTMLInputElement).value = str;
		doFactor(null);
	}
	
	
	
	class GaussianInteger {
		
		constructor(
			public readonly real: number,
			public readonly imag: number) {}
		
		
		private norm(): number {
			return this.real * this.real + this.imag * this.imag;
		}
		
		
		private multiply(other: GaussianInteger): GaussianInteger {
			return new GaussianInteger(
				this.real * other.real - this.imag * other.imag,
				this.real * other.imag + this.imag * other.real);
		}
		
		
		private isDivisibleBy(re: number, im: number): boolean {
			const divisorNorm: number = re * re + im * im;
			return ( this.real * re + this.imag * im) % divisorNorm == 0 &&
			       (-this.real * im + this.imag * re) % divisorNorm == 0;
		}
		
		
		private divide(other: GaussianInteger): GaussianInteger {
			if (!this.isDivisibleBy(other.real, other.imag))
				throw "Cannot divide";
			return new GaussianInteger(
				( this.real * other.real + this.imag * other.imag) / other.norm(),
				(-this.real * other.imag + this.imag * other.real) / other.norm());
		}
		
		
		public factorize(): Array<GaussianInteger> {
			if (this.norm() <= 1)  // 0, 1, -1, i, -i
				return [this];
			
			let result: Array<GaussianInteger> = [];
			let temp: GaussianInteger = this;
			let check = new GaussianInteger(1, 0);
			while (temp.norm() > 1) {
				const factor = temp.findPrimeFactor();
				result.push(factor);
				temp = temp.divide(factor);
				check = check.multiply(factor);
			}
			check = check.multiply(temp);
			if (temp.norm() != 1 || check.real != this.real || check.imag != this.imag)
				throw "Assertion error";
			if (temp.real != 1)  // -1, i, -i
				result.push(temp);
			
			result.sort((x, y) => {
				if      (x.norm() < y.norm()) return -1;
				else if (x.norm() > y.norm()) return +1;
				else if (x.real > y.real) return -1;
				else if (x.real < y.real) return +1;
				else return 0;
			});
			return result;
		}
		
		
		private findPrimeFactor(): GaussianInteger {
			const norm: number = this.norm();
			if (norm % 2 == 0)
				return new GaussianInteger(1, 1);
			
			for (let i = 3, end = Math.floor(Math.sqrt(norm)); i <= end; i += 2) {  // Find factors of norm
				if (norm % i == 0) {
					if (i % 4 == 3)
						return new GaussianInteger(i, 0);
					else {
						for (let re = Math.floor(Math.sqrt(i)); re > 0; re--) {
							const im = Math.round(Math.sqrt(i - re * re));
							if (re * re + im * im == i && this.isDivisibleBy(re, im))
								return new GaussianInteger(re, im);
						}
					}
				}
			}
			
			// This number itself is prime. Rotate so that the argument is in [0, pi/2)
			let temp: GaussianInteger = this;
			while (temp.real < 0 || temp.imag < 0)
				temp = temp.multiply(new GaussianInteger(0, 1));
			return temp;
		}
		
		
		public toString(): string {
			if (this.real == 0 && this.imag == 0)
				return "0";
			else {
				let result: string = "";
				if (this.real != 0)
					result += this.real > 0 ? this.real : MINUS + (-this.real);
				if (this.imag != 0) {
					if (result == "")
						result += this.imag > 0 ? "" : MINUS;
					else
						result += this.imag > 0 ? " + " : ` ${MINUS} `;
					result += (Math.abs(this.imag) != 1 ? Math.abs(this.imag) : "") + "i";
				}
				return result;
			}
		}
		
		
		public static parseString(str: string): GaussianInteger {
			if (/\d\s+\d/.test(str))  // Spaces are not allowed between digits
				throw "Invalid number";
			str = str.replace(/\s+/g, "");  // Remove all whitespace
			str = str.replace(/\u2212/g, "-");
			str = str.replace(/j/g, "i");
			
			function checkedParseInt(s: string): number {
				const n = parseInt(s, 10);
				if (Math.abs(n) >= 67108864)
					throw "Number is too large";
				return n;
			}
			
			// Match one of the syntax cases
			let real: number, imag: number;
			let mat: RegExpExecArray|null;
			if ((mat = /^([+-]?\d+)$/.exec(str)) !== null) {  // e.g. 1, +0, -2
				real = checkedParseInt(mat[1]);
				imag = 0;
			} else if ((mat = /^([+-]?)(\d*)i$/.exec(str)) !== null) {  // e.g. i, 4i, -3i
				real = 0;
				imag = checkedParseInt(mat[1] + (mat[2] != "" ? mat[2] : "1"));
			} else if ((mat = /^([+-]?\d+)([+-])(\d*)i$/.exec(str)) !== null) {  // e.g. 1+2i, -3-4i, +5+i
				real = checkedParseInt(mat[1]);
				imag = checkedParseInt(mat[2] + (mat[3] != "" ? mat[3] : "1"));
			} else if ((mat = /^([+-]?)(\d*)i([+-]\d+)$/.exec(str)) !== null) {  // e.g. 2i+1, -4i-3, +i+5
				real = checkedParseInt(mat[3]);
				imag = checkedParseInt(mat[1] + (mat[2] != "" ? mat[2] : "1"));
			} else
				throw "Invalid number";
			return new GaussianInteger(real, imag);
		}
		
	}
	
	
	const MINUS = "\u2212";
	const NBSP  = "\u00A0";
	
}
